#Week1's Homework Machine Learning Zoomcamp

by Rahma Hayuning Astuti


### Question 1

What's the version of Pandas that you installed?

In [58]:
import pandas as pd
pd.__version__

'1.3.4'

So, the version of Pandas that I installed is 1.3.4

### Question 2

How many columns are in the dataset?

- 10
- 6560
- 10989
- 20640

In [59]:
data = pd.read_csv('D:\MLZoomcamp\Homework\Week 1\housing.csv') #read file
data

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

In [60]:
data.shape

(20640, 10)

It shows that both using 'read_csv' and 'shape' results are the dataset has 10 column. So, the answer is A or the first bullet. 

### Question 3

Which columns in the dataset have missing values?

- `total_rooms`
- `total_bedrooms`
- both of the above
- no empty columns in the dataset

In [61]:
data.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [62]:
data.isnull().any()

longitude             False
latitude              False
housing_median_age    False
total_rooms           False
total_bedrooms         True
population            False
households            False
median_income         False
median_house_value    False
ocean_proximity       False
dtype: bool

from using isnull().sum() and isnull.any(), can conclude that the columns in the dataset have missing values is 'total_bedrooms'

### Question 4

How many unique values does the `ocean_proximity` column have?

- 3
- 5
- 7
- 9

In [63]:
data['ocean_proximity'].nunique()

5

'ocean_proximity' column has 5 unique value

### Question 5

What's the average value of the `median_house_value` for the houses located near the bay?

- 49433
- 124805
- 259212
- 380440

In [64]:
avg = data.groupby('ocean_proximity')['median_house_value'].mean()
print(avg)

ocean_proximity
<1H OCEAN     240084.285464
INLAND        124805.392001
ISLAND        380440.000000
NEAR BAY      259212.311790
NEAR OCEAN    249433.977427
Name: median_house_value, dtype: float64


In [65]:
avg.round(decimals=0)

ocean_proximity
<1H OCEAN     240084.0
INLAND        124805.0
ISLAND        380440.0
NEAR BAY      259212.0
NEAR OCEAN    249434.0
Name: median_house_value, dtype: float64

the average value of 'median_house_value' that based on the house located near the bay is 259212.311790. So, the answer is 259212

### Question 6

1. Calculate the average of `total_bedrooms` column in the dataset.
2. Use the `fillna` method to fill the missing values in `total_bedrooms` with the mean value from the previous step.
3. Now, calculate the average of `total_bedrooms` again.
4. Has it changed?

Has it changed?

In [66]:
#step 1 find the average value
avg_total_bedrooms = data['total_bedrooms'].mean().round(decimals= 3)
avg_total_bedrooms

537.871

In [67]:
#step 2 fill the empty data
fill = data['total_bedrooms'].fillna(avg_total_bedrooms)

In [68]:
#check if data are filled
fill.isnull().sum()

0

In [69]:
#step3 average
fill.mean().round(decimals= 3)

537.871

In [70]:
fill.isna().sum()

0

based on the previous calculated and process before, the average of `total_bedrooms` doesn't change. So, the answer is No

### Question 7

1. Select all the options located on islands.
2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the value of the last element of `w`?

- -1.4812
- 0.001
- 5.6992
- 23.1233

In [71]:
#step 1 -3
X = data[['housing_median_age', 'total_rooms', 'total_bedrooms']] [data['ocean_proximity'] == "ISLAND"]

In [72]:
print(X)

      housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0


In [73]:
#shape of X
X.shape

(5, 3)

In [74]:
#shape of X transpose
XT = X.transpose()
XT.shape

(3, 5)

In [76]:
#result of XTX
XTX = XT.dot(X)
print(XTX)

                    housing_median_age  total_rooms  total_bedrooms
housing_median_age              9682.0     351053.0         91357.0
total_rooms                   351053.0   14399307.0       3772036.0
total_bedrooms                 91357.0    3772036.0        998358.0


In [77]:
XTX.shape

(3, 3)

In [78]:
#inverse XTX
import numpy as np
XTX_inv  = np.linalg.inv(XTX)
print(XTX_inv)

[[ 9.19403586e-04 -3.66412216e-05  5.43072261e-05]
 [-3.66412216e-05  8.23303633e-06 -2.77534485e-05]
 [ 5.43072261e-05 -2.77534485e-05  1.00891325e-04]]


In [79]:
#step 6
y = [950, 1300, 800, 1000, 1300]

In [80]:
#step 7
a = XTX_inv.dot(XT)
w= a.dot(y)
w


array([23.12330961, -1.48124183,  5.69922946])

In [82]:
w.round(decimals = 4)


array([23.1233, -1.4812,  5.6992])

The last value of w is 5.6992